File: leetcode_go/greedy_algorithms/gfg_minimum_platforms.go
/**
 * GFG: Minimum Platforms
 * Link: https://www.geeksforgeeks.org/problems/minimum-platforms-1587115620/1
*/

package main

import (
	"fmt"
	"sort"
)

func minPlatforms(arr []int, dep []int)	int {
	sort.Ints(arr[:])
	sort.Ints(dep[:])

	i := 0
	j := 0
	cnt := 0
	maxCnt := 0

	for i < len(arr) {
		if arr[i] <= dep[j] {
			cnt += 1
			i += 1
		} else {
			cnt -= 1
			j += 1
		}
		maxCnt = max(maxCnt, cnt)
	}
	return maxCnt
}

func main(){
	arr := []int {900,945,955,1100,1500,1800}
	dep := []int {920,1200,1130,1155,1900,2000}
	
	fmt.Println(minPlatforms(arr, dep))
}

File: leetcode_rs/greedy_algorithms/gfg_minimum_platforms.rs
/**
 * GFG: Minimum Platforms
 * Link: https://www.geeksforgeeks.org/problems/minimum-platforms-1587115620/1
*/

fn min_platforms(mut arr: Vec<i32>, mut dep: Vec<i32>) -> i32 {
    arr.sort();
    dep.sort();

    let mut i: usize = 0;
    let mut j: usize = 0;
    let mut cnt: i32 = 0;
    let mut max_cnt: i32 = 0;

    while i < arr.len() {
        if arr[i] < dep[j] {
            cnt+=1;
            i += 1;
        } else {
            cnt -= 1;
            j += 1;
        }
        max_cnt = std::cmp::max(cnt, max_cnt);
    }
    max_cnt
}

fn main(){
    let arr: Vec<i32> = vec![900, 945, 955, 1100, 1500, 1800];
    let dep: Vec<i32> = vec![920, 1200, 1130, 1150, 1900, 2000];

    println!("{}", min_platforms(arr, dep));
}